### PR TITLE
Support for Heap after GC stats

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
JVM stats published today represents the point in time usage of the heap across pools. While
`MemoryPoolMXBean::getCollectionUsage` provides the memory usage after the Java virtual machine most recently expended effort in recycling unused objects in this memory pool. Exposing this information could be quite helpful while troubleshooting GC and memory issues.

The `jvm` section of the node statistics had been extended to include the additional details regarding most recent GC cycle for a particular memory pool, under new `last_gc_stats` section:
```
            "old" : {                             
              "used_in_bytes" : 50526632,
              "max_in_bytes" : 357957632,
              "peak_used_in_bytes" : 50526632,
              "peak_max_in_bytes" : 357957632,
              "last_gc_stats" : {
                "used_in_bytes" : 42130040,
                "max_in_bytes" : 357957632,
                "usage_percent" : 11
              }                                   
            }  
```

The sample output:

```
curl http://localhost:9200/_nodes/stats/jvm?pretty
```

```
....
{
        ...
        "pools" : {                                                                                                                                                                                      
            "young" : {                                                                                                                                                                                    
              "used_in_bytes" : 103327560,                                                                                                                                                                 
              "max_in_bytes" : 143130624,                                                                                                                                                                  
              "peak_used_in_bytes" : 143130624,                                                                                                                                                            
              "peak_max_in_bytes" : 143130624,                                                                                                                                                             
              "last_gc_stats" : {                                                                                                                                                                          
                "used_in_bytes" : 73248760,                                                                                                                                                                
                "max_in_bytes" : 143130624,                                                                                                                                                                
                "usage_percent" : 51              
              }                                   
            },                                    
            "survivor" : {                        
              "used_in_bytes" : 17431960,         
              "max_in_bytes" : 17891328,          
              "peak_used_in_bytes" : 17891328,    
              "peak_max_in_bytes" : 17891328,     
              "last_gc_stats" : {          
                "used_in_bytes" : 17431960,       
                "max_in_bytes" : 17891328,
                "usage_percent" : 97              
              }                                   
            },                                    
            "old" : {                             
              "used_in_bytes" : 50526632,
              "max_in_bytes" : 357957632,
              "peak_used_in_bytes" : 50526632,
              "peak_max_in_bytes" : 357957632,
              "last_gc_stats" : {
                "used_in_bytes" : 42130040,
                "max_in_bytes" : 357957632,
                "usage_percent" : 11
              }                                   
            }                                     
          }                                       
        },  
        ...
```

 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/981
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@Bukhtawar may I ask you please to take a look, is it something you had mind or there are missing pieces? thank you!

**Review by dblock — APPROVED** [review]
LGTM

**Comment by dblock** [comment]
start grade check

File: server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java [hunk 1]
```diff
@@ -478,7 +557,8 @@ public ByteSizeValue getPeakMax() {
         private final long nonHeapUsed;
         private final List<MemoryPool> pools;
 
-        public Mem(long heapCommitted, long heapUsed, long heapMax, long nonHeapCommitted, long nonHeapUsed, List<MemoryPool> pools) {
+        public Mem(long heapCommitted, long heapUsed, long heapMax, long nonHeapCommitted, long nonHeapUsed, 
+                List<MemoryPool> pools) {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:561]: Unrelated formatting?

File: server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java [hunk 2]
```diff
@@ -414,6 +433,49 @@ public int getPeakCount() {
...
+
+        public ByteSizeValue getMax() {
+            return new ByteSizeValue(max);
+        }
+        
+        public short getUsagePercent() {
+            if (max == 0) {
+                return -1;
+            }
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:475]: nit: extra line break

File: server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java [hunk 3]
```diff
@@ -299,7 +316,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         static final String PEAK_USED_IN_BYTES = "peak_used_in_bytes";
         static final String PEAK_MAX = "peak_max";
         static final String PEAK_MAX_IN_BYTES = "peak_max_in_bytes";
-
+        static final String USAGE_PERCENT = "usage_percent";
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:319]: Can this be confused with current usage vs usage after GC?

Comment (reta) [reply server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:319]: Good question, I think it should not since the field is under `last_gc_stats` object, in this context, it should not be confusing. Also, the semantic of this field stays consistent between current usage and usage after GC.  What do you think?

File: server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java (hunk 1 above)
Comment (reta) [reply server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:561]: Fixed, thanks!

File: server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java (hunk 2 above)
Comment (reta) [reply server/src/main/java/org/opensearch/monitor/jvm/JvmStats.java:475]: Fixed, thanks!

**Comment by reta** [comment]
@Bukhtawar anything pending for this change from your side? thank you! :-)

**Review by Bukhtawar — APPROVED** [review]
LGTM.. Thanks @reta for the changes

**Comment by reta** [comment]
@dblock may I ask you please to schedule Gradle check? Thank you!

**Comment by dblock** [comment]
start gradle check

**Comment by reta** [comment]
>  Gradle Check failure [2c362e4](https://github.com/opensearch-project/OpenSearch/commit/2c362e40ab4ed1d3fa3af6b7fd00490fab337c1b)
> [Log 569](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_569.log)
> 
> [Reports 569](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_569_reports.zip)

Hm ... `Task :distribution:bwc:minor:buildBwcLinuxTar FAILED`

```
> Task :distribution:bwc:staged:buildBwcLinuxTar
 [1.1.0] > Task :distribution:archives:linux-tar:assemble
 [1.1.0] build complete, generating: /var/CITOOL/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/search/distribution/bwc/staged/build/bwc/checkout-1.1/build/569.tar.bz2

> Task :distribution:bwc:minor:buildBwcLinuxTar FAILED
> Task :modules:repository-url:urlFixture#stop

> Task :distribution:bwc:staged:buildBwcLinuxTar
 [1.1.0] 
 [1.1.0] BUILD SUCCESSFUL in 6m 32s
 [1.1.0] 149 actionable tasks: 149 executed

> Task :distribution:bwc:staged:buildBwcLinuxTar FAILED
```

Same as https://github.com/opensearch-project/OpenSearch/issues/1285 ?

**Comment by reta** [comment]
My apologies, @dblock , seems like  https://github.com/opensearch-project/OpenSearch/issues/1285 was closed, may I ask you please to reschedule the Gradle check whenever you have time, thank you!

**Comment by dblock** [comment]
start gradle check

**Comment by dblock** [comment]
Thanks for hanging in here! I kicked it, but I think this might need a rebase.

**Comment by reta** [comment]
Rebased, thanks a lot @dblock !

**Comment by dblock** [comment]
start gradle check

**Comment by reta** [comment]
> x   Gradle Check failure [e7155ca](https://github.com/opensearch-project/OpenSearch/commit/e7155cad1b404c284423c125df29d82e1e85c442) [Log 586](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_586.log)
> 
> [Reports 586](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_586_reports.zip)

On it

**Comment by dblock** [comment]
start gradle check
